File: cpp/monte_carlo_single.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <ctime>
#include "xoshiro256.hpp"

// 理論上の円周率
const double PI_THEORETICAL = 3.141592653589793238462643383279502884197;

/**
 * モンテカルロ法で円周率を計算
 * 
 * @param iterations 試行回数
 * @param pi_estimate 円周率の推定値（出力）
 * @param error 誤差（出力）
 */
void calculate_pi(uint64_t iterations, double &pi_estimate, double &error) {
    Xoshiro256 rng(12345);  // 固定シード
    
    uint64_t inside_circle = 0;
    
    for (uint64_t i = 0; i < iterations; i++) {
        // 0.0以上1.0未満の乱数を生成
        double x = rng.next_double();
        double y = rng.next_double();
        
        // 単位円内か判定（x^2 + y^2 <= 1）
        if (x * x + y * y <= 1.0) {
            inside_circle++;
        }
    }
    
    // π ≈ 4 × (円内の点数) / (総試行回数)
    pi_estimate = 4.0 * inside_circle / iterations;
    error = std::abs(pi_estimate - PI_THEORETICAL);
}

int main() {
    const uint64_t iterations = 100000000ULL;
    
    // 実行時間の測定
    std::clock_t start = std::clock();
    double pi_estimate, error;
    calculate_pi(iterations, pi_estimate, error);
    std::clock_t end = std::clock();
    
    double elapsed_ms = ((double)(end - start) / CLOCKS_PER_SEC) * 1000.0;
    
    // 結果をJSON形式で出力
    std::cout << std::fixed << std::setprecision(15);
    std::cout << "{\n";
    std::cout << "  \"language\": \"C++\",\n";
    std::cout << "  \"variant\": \"standard\",\n";
    std::cout << "  \"version\": \"C++17\",\n";
    std::cout << "  \"mode\": \"single\",\n";
    std::cout << "  \"iterations\": " << iterations << ",\n";
    std::cout << "  \"pi_estimate\": " << pi_estimate << ",\n";
    std::cout << "  \"error\": " << error << ",\n";
    std::cout << std::setprecision(2);
    std::cout << "  \"time_ms\": " << elapsed_ms << ",\n";
    std::cout << "  \"memory_mb\": 0.0,\n";  // runner.pyで測定
    std::cout << "  \"cache_misses\": 0,\n";  // runner.pyで測定
    std::cout << "  \"lines_of_code\": 0,\n";  // runner.pyで計算
    std::cout << "  \"compiler_flags\": \"-O3 -march=native -flto\",\n";
    std::cout << "  \"cpu_model\": \"N/A\",\n";
    std::cout << "  \"cpu_cores\": 1,\n";
    std::cout << "  \"thread_count\": 1,\n";
    std::cout << "  \"os\": \"N/A\",\n";
    std::cout << "  \"os_version\": \"N/A\",\n";
    std::cout << "  \"compiler\": \"GCC/Clang\",\n";
    std::cout << "  \"simd_detected\": false,\n";
    std::cout << "  \"simd_instructions\": []\n";
    std::cout << "}\n";
    
    return 0;
}

File: fortran/monte_carlo_single.f90
program monte_carlo_single
    use xoshiro256_module
    implicit none
    
    real(8), parameter :: PI_THEORETICAL = 3.141592653589793238462643383279502884197d0
    integer(8), parameter :: iterations = 100000000_8
    type(xoshiro256_type) :: rng
    integer(8) :: i, inside_circle
    real(8) :: x, y, pi_estimate, error
    real(8) :: start_time, end_time, elapsed_ms
    
    call cpu_time(start_time)
    
    call xoshiro256_init(rng, 12345_8)
    inside_circle = 0
    
    do i = 1, iterations
        x = xoshiro256_next_double(rng)
        y = xoshiro256_next_double(rng)
        
        if (x * x + y * y <= 1.0d0) then
            inside_circle = inside_circle + 1
        end if
    end do
    
    pi_estimate = 4.0d0 * real(inside_circle, 8) / real(iterations, 8)
    error = abs(pi_estimate - PI_THEORETICAL)
    
    call cpu_time(end_time)
    elapsed_ms = (end_time - start_time) * 1000.0d0
    
    write(*, '(a)') '{'
    write(*, '(a)') '  "language": "Fortran",'
    write(*, '(a)') '  "variant": "standard",'
    write(*, '(a)') '  "version": "F2008",'
    write(*, '(a)') '  "mode": "single",'
    write(*, '(a,i0,a)') '  "iterations": ', iterations, ','
    write(*, '(a,f0.15,a)') '  "pi_estimate": ', pi_estimate, ','
    write(*, '(a,f0.15,a)') '  "error": ', error, ','
    write(*, '(a,f0.2,a)') '  "time_ms": ', elapsed_ms, ','
    write(*, '(a)') '  "memory_mb": 0.0,'
    write(*, '(a)') '  "cache_misses": 0,'
    write(*, '(a)') '  "lines_of_code": 0,'
    write(*, '(a)') '  "compiler_flags": "-O3 -march=native -flto",'
    write(*, '(a)') '  "cpu_model": "N/A",'
    write(*, '(a)') '  "cpu_cores": 1,'
    write(*, '(a)') '  "thread_count": 1,'
    write(*, '(a)') '  "os": "N/A",'
    write(*, '(a)') '  "os_version": "N/A",'
    write(*, '(a)') '  "compiler": "GCC/GFortran",'
    write(*, '(a)') '  "simd_detected": false,'
    write(*, '(a)') '  "simd_instructions": []'
    write(*, '(a)') '}'
    
end program monte_carlo_single

